packages/amplify-graphql-model-transformer/src/resources/amplify-dynamodb-table/amplify-table-manager-lambda/amplify-table-manager-handler.ts

```diff
@@ -726,6 +733,24 @@ export const extractTableInputFromEvent = (
   return tableDef;
 };
 
+/**
+ * Extract the old custom DynamoDB table properties from event, during which the service token will be removed
+ * and the string values will be correctly parsed to boolean or number
+ * @param event Event for onEvent or isComplete
+ * @returns The old table input for Custom dynamoDB Table
+ */
+export const extractOldTableInputFromEvent = (
+  event: AWSCDKAsyncCustomResource.OnEventRequest | AWSCDKAsyncCustomResource.IsCompleteRequest,
+): CustomDDB.Input => {
+  // isolate the resource properties from the event and remove the service token
+  const resourceProperties = { ...event.OldResourceProperties } as Record<string, any> & { ServiceToken?: string };
+  delete resourceProperties.ServiceToken;
+
+  // cast the remaining resource properties to the DynamoDB API call input type
+  const tableDef = convertStringToBooleanOrNumber(resourceProperties) as CustomDDB.Input;
```

Comment:
A little bit of a tangent, but I'm unclear why we're casting this to CustomDDB.Input. If I had to guess, I'd say that the incoming event is emitted by CDK, and the `OldResourceProperties` represented the event that last created/updated the resource. Since we know that the resource was created using our handler, we can cast it as a `CustomDDB.Input`. Is that right?

If that's the case, how are we ensuring that we maintain this input contract? Is the `CustomDDB.Input` API documented as public and locked so we don't accidentally break existing installations by changing a field name or adding a new required field that old installations might not have?

Reply:
I've finally use the combination of `maxRetry=5` and `delay=5s`. Some datapoints are as follows.
| max retry | initial delay (s) | total deployment time (s)|
|---|---|---|
|5|10|407.1|
|5|0|383.82|
|5|5|382.45|

Reply:
Thanks. Please add a doc comment in the CustomDDB.Input type that this type defines how data is stored in CloudFormation, so changes to the structure must be backward compatible with all previous versions.

Reply:
Added in [83c56ae](https://github.com/aws-amplify/amplify-category-api/pull/2410/commits/83c56aebf1507b757b28b981f741fc1030fb2a0a)